NMS-12373: Provisiond Docs Update (#3521)
OpenNMS/opennms: docs/modules/operation/pages/provisioning/import-handler.adoc

[[import-handlers]]
= Import handlers

{page-component-title} supports a number of import handlers to facilitate provisioning nodes from an external system (foreign source) into a requisition.
This is done in the `provisiond-configuration.xml` file by referencing a requisition's location via a URL.
Most commonly, these URLs are files on the file system (`file:/opt/opennms/etc/imports/<my-requisition.xml>`) created through the Requisitions UI.
However, these same requisitions for adding, updating, and deleting nodes can also come from external URLs.
For example, a requisition can be retrieved using the HTTP protocol.

In addition to the standard protocols supported by Java, we provide a series of custom URL handlers to help retrieve requisitions from external sources.

== Generic handler

The generic handler is made available using URLs of the form: `requisition://type?param=1;param=2`

> Using these URLs, various type handlers can be invoked, both locally and via a Minion.

Reviewer: ```suggestion
Using these URLs, various types of handlers can be invoked, both locally and via a Minion.
```

In addition to the type-specific parameters, the following parameters are supported by all handlers:

.Common parameters
[options="header, autowidth"]
|===
| Parameter              | Description                                                                    | Required | Default value

> | `location`             | The name of location at which the handler should be run                        | optional | Default

Reviewer: ```suggestion
| `location`             | The name of location at which the handler should run.                        | optional | Default
```

> | `ttl`                  | The maximum number of milliseconds to wait for the handler when ran remotely   | optional | 20000

Reviewer: ```suggestion
| `ttl`                  | The maximum number of milliseconds to wait for the handler when run remotely   | optional | 20000
```

Author: ```suggestion
| `ttl`                  | The maximum number of milliseconds to wait for the handler when run remotely.  | optional | 20000
```

|===

See the relevant sections below for additional details on the supported types.

> The `opennms:show-import` command available via the Karaf Shell can be used to show the results of an import without persisting or triggering the import:

Reviewer: ```suggestion
Use the `opennms:show-import` command available via the Karaf Shell to show the results of an import without persisting or triggering the import:
```

The generic handler is made available using URLs of the form: `requisition://type?param=1;param=2`

> Using these URLs, various types of handlers can be invoked, both locally and via a Minion.

Reviewer: test

In addition to the type-specific parameters, the following parameters are supported by all handlers:

.Common parameters
[options="header, autowidth"]
|===
| Parameter              | Description                                                                    | Required | Default value

> | `location`             | The name of location at which the handler should run.                        | optional | Default

Reviewer: ```suggestion
| `location`             | The name of the location at which the handler should run.                        | optional | Default
```

| `ttl`                  | The maximum number of milliseconds to wait for the handler when run remotely.  | optional | 20000
|===

See the relevant sections below for additional details on the supported types.

> Use the `opennms:show-import` command available via the Karaf Shell to show the results of an import without persisting or triggering the import:

Reviewer: ```suggestion
Use the `opennms:show-import` command, available via the Karaf shell, to show the results of an import without persisting or triggering the import:
```

[source]
----
http://myinventory.server.org/nodes.cgi
----

=== Using the generic handler

[source]
----
requisition://http?url=http%3A%2F%2Fmyinventory.server.org%2Fnodes.cgi
----

TIP: When using the generic handler, the URL should be "URL encoded".

=== Using the OpenNMS Provisioning Integration Server (PRIS)

[source]
----
http://localhost:8000/requisitions/myServer
----

== DNS handler

The DNS handler requests a Zone Transfer (AXFR) from a DNS server.
The A and AAAA records are retrieved and used to build an import requisition.
This is handy for organizations that use DNS, possibly coupled with an IP management tool, as the database of record for nodes in the network.
So, rather than ping sweeping the network or entering the nodes manually into the {page-component-title} Requisition UI, nodes can be managed via one or more DNS servers.

The format of the URL for this new protocol handler is: `dns://<host>[:port]/<zone>[/<foreign-source>/][?expression=<regex>]`

DNS Import Examples:

=== Simple

[source]
----
dns://my-dns-server/myzone.com
----

This URL will import all A and AAAA records from the zone `myzone.com` on the host `my-dns-server` via port 53 (default DNS port).

> Since the foreign source is not specified it will add to the default requisition.

Reviewer: ```suggestion
Since the foreign source is not specified, it will be added to the default requisition.
```

----

== Location

You can use the `location` parameter to assign a single location to all nodes in this requisition.
This would only apply if you are using Minions in your network.

[source]
----
dns://my-dns-server/myzone.com/portland/?location=Raleigh
----

By prefixing with a `~`, you can specify a url-encoded regular expression to determine the location based on the host entries.
The following example derives the location based on the subdomain.
This example uses a URL-encoded regular expression `^(?:.*\.|)(.*?)\.hs-fulda\.de\.$` with exactly one capturing group containing the subdomain.

[source]
----
dns://my-dns-server/hs-fulda.de/?location=~%5E%28%3F%3A.%2A%5C.%7C%29%28.%2A%3F%29%5C.hs-fulda%5C.de%5C.%24
----

This will associate nodes for the host entries `foo.e46.hs-fulda.de.` and `bar.e46.hs-fulda.de.` to the location `e46` while hosts like `aaa.g51.hs-fulda.de` and `bbb.g51.hs-fulda.de` will be assigned location `g51`.

TIP: You can use online tools like https://www.urlencoder.org to encode your parameters.

=== DNS setup

The DNS server must be configured to allow a zone transfer from the {page-component-title} server.

> // Question: Is this still recommended to setup a local DNS secondary?

Reviewer: ```
// Question: Is this still recommended to setup a local DNS secondary?
```
I am not sure. Will have to ask someone who might know.

Reviewer 2: We can remove this comment / TODO here. You have to cover security issues when you allow zone transfers between your OpenNMS and the DNS server. For us, as OpenNMS, it's first a matter of fact we need to access the DNS server using AXFR/IXFR. IMHO, what strategies you need to address this topic of security in your DNS environment is a different story.

Author: ```suggestion
```

=== DNS setup

The DNS import handler performs an AXFR zone transfer request to collect DNS resource records for use in provisioning nodes.
Zone transfers are designed for zone replication between DNS servers.
You may want to provide additional security mechanisms to protect your DNS infrastructure.

> At the minimum, you should always restrict AXFR requests to only trusted IP addresses and/or use a hidden, read-only secondary DNS servers for {page-component-title}.

Reviewer: ```suggestion
At a minimum, you should always restrict AXFR requests to only trusted IP addresses and/or use a hidden, read-only secondary DNS server for {page-component-title}.
```

> One suggestion is to setup a secondary DNS server on your {page-component-title} and configure the rest of the DNS servers to allow transfers to your {page-component-title} server.

Reviewer: ```suggestion
One suggestion is to set up a secondary DNS server on your {page-component-title} and configure the rest of the DNS servers to allow transfers to your {page-component-title} server.
```

> This will allow you to point the DNS import handler to `localhost`.

Reviewer: ```suggestion
This allows you to point the DNS import handler to `localhost`.
```
